CRD backed Identities in K8s

raybejjani opened the pull request:
This PR follows on from https://github.com/cilium/cilium/pull/8496. It looks large but most of it is auto-generated code. The two code parts are the new CRD in `pkg/k8s/apis/cilium.io/v2/types.go` and the identity backend in `pkg/k8s/identitybackend/identity.go`. The rest is boilerplate or code to support the CRD (such as gargbage collection in cilium-operator in `operator/k8s_identity.go`).

The new ciliumidentity CRD acts as the backing store for identity allocation in cilium (generic backends were introduced in #8496). Most pkg/allocator.Backend operations are supported but, specifically, locking is not.

The name of the CRD is the numeric identity and the labels on the CRD object are the security labels used by cilium when assigning the numeric identity. Every node using a particular ciliumidentity updates the `status` field with it's name and a timestamp. cilium-operator can then use the list of nodes in the `status` field to reference count users (nodes) of this identity, and to expire stale usage.

Both cilium and cilium-operator need full access to create, update and delete these entities. cilium-operator does the garbage collection on these objects and can also synchronise them to a KVStore.

A CRD example:
```
 {
            "apiVersion": "cilium.io/v2",
            "kind": "CiliumIdentity",
            "metadata": {
                "creationTimestamp": "2019-07-16T00:53:58Z",
                "generation": 1,
                "labels": {
                    "class": "deathstar",
                    "io.cilium.k8s.policy.cluster": "default",
                    "io.cilium.k8s.policy.serviceaccount": "default",
                    "io.kubernetes.pod.namespace": "default",
                    "org": "empire"
                },
                "name": "63315",
                "resourceVersion": "25518",
                "selfLink": "/apis/cilium.io/v2/ciliumidentities/63315",
                "uid": "43337e9a-f510-4438-aa69-c65025d4aad2"
            },
            "securitylabels": {
                "k8s:class": "deathstar",
                "k8s:io.cilium.k8s.policy.cluster": "default",
                "k8s:io.cilium.k8s.policy.serviceaccount": "default",
                "k8s:io.kubernetes.pod.namespace": "default",
                "k8s:org": "empire"
            }
        }
```


2019-07-15
- Switched ciliumid CRD to be namespace-free
- removed `identities` as a CRD name for ciliumid
- removed --synchronize-identities option in cilium-operator because it wasn't actually implemented.
- explicit check for allocation-mode options that aren’t supported 
- debug print on failed status updates in k8s identity allocator
- simplified requiresKVStore function in operator
- removed unused onAdd/DeleteIdentity handlers in operator
- switched startCRDIdentityGC to be a controller
- use defaults.KVStoreLeaseTTL as the default identityGCInterval value
- clearer debug and comment when falling through during status updates


<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8236)
<!-- Reviewable:end -->

tgraf commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -458,6 +458,35 @@ type EndpointIdentity struct {
 	LabelsSHA256 string `json:"labelsSHA256,omitempty"`
 }
 
+// +genclient
```
Add:
```
// +genclient:nonNamespaced
```

To not require to specify the namespace in client code

raybejjani replied:
done! excellent

raybejjani committed:
k8s: Move k8s/informer benchmarks to k8s/informer/benchmarks package

The informer benchmarks need the k8s package to initialize. This can
cause an import loop when running tests and other packages that rely on
informer are also used by k8s. This occurswhen their tests also need to
initialize using pkg/k8s, often to obtain the k8s client.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
k8s: Add ciliumidentity CRD

CiliumIdentity is a CRD that represents an identity managed by Cilium.
It is intended as a backing store for identity allocation, acting as the
global coordination backend, and can be used in place of a KVStore (such
as etcd).
The name of the CRD is the numeric identity and the labels used for the
identity are in the security-labels field. The labels on the CRD object
are a filtered subset.
Each node using a ciliumidentity updates the status field with it's name
and a timestamp.  cilium-operator can then use the list of nodes in
status to reference count users of this identity, and to expire stale
usage.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
k8s: Add RBAC for k8s CRD cilium identities

The new ciliumidentity CRD is a new type and RBAC needs to be updated
for it. Both cilium and cilium-operator need full access to create,
update and delete these entities. cilium-operator does the garbage
collection on these objects and can also syncronize them to a KVStore.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
k8s: Add CRD Identities as an identity allocator backend

The ciliumidentity CRD can act as the backing store for identity
allocation in cilium. Implementation details are hidden behind the
pkg/allocator.Backend interface. Most operations are supported but,
specifically, locking is not.

The name of the CRD is the numeric identity and the labels on the CRD
object are the security labels used by cilium. Each node using a
ciliumidentity updates the status field with it's name and a timestamp.
cilium-operator can then use the list of nodes in status to reference
count users of this identity, and to expire stale usage.

Both cilium and cilium-operator need full access to create, update and
delete these entities. cilium-operator does the garbage collection on
these objects.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -287,6 +287,32 @@ func createNodeCRD(clientset apiextensionsclient.Interface) error {
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural:     "ciliumidentities",
+				Singular:   "ciliumidentity",
+				ShortNames: []string{"ciliumid", "identity"},
```
I would remove `"identity"` here, it seems too generic.

raybejjani replied:
good point, any other ideas? `ciliumid` feels a little wrong still but I can't think of anything else.

nebril replied:
`securityID`?

aanm reviewed (commented):
I'll review it later

tgraf commented on operator/k8s_identity.go:
```diff
@@ -0,0 +1,123 @@
+			}
+		}
+
+		log.Infof("Deleting unused identity %+v", identity)
```
Looks like a leftover from my original code. I think we should just print name and not the full structure.

raybejjani replied:
I made this a debug since it might be useful (and doesn't happen too often).

tgraf commented on operator/k8s_identity.go:
```diff
@@ -0,0 +1,123 @@
+	)
+
+	go identityInformer.Run(wait.NeverStop)
+	log.Info("Starting to synchronize CRD backed identities to kvstore")
```
Synchronization to kvstore is not added yet so we should change this info message

raybejjani replied:
I haven't pushed yet to allow tests to run, but I'll pull the synchronisation option into another PR since I need to implement it :/

tgraf commented on operator/main.go:
```diff
@@ -112,8 +115,11 @@ func init() {
 	flags.StringVar(&metricsAddress, "metrics-address", ":6942", "Address to serve Prometheus metrics")
 	flags.BoolVar(&synchronizeServices, "synchronize-k8s-services", true, "Synchronize Kubernetes services to kvstore")
 	flags.BoolVar(&synchronizeNodes, "synchronize-k8s-nodes", true, "Synchronize Kubernetes nodes to kvstore and perform CNP GC")
+	flags.BoolVar(&synchronizeIdentities, "synchronize-identities", true, "Synchronize CRD backed identities to kvstore")
```
We should enable garbage collector if identity allocation is set to crd and add the `synchronize-identities` flag once we add the synchronization functionality

raybejjani replied:
oh, and it currently enables CRD GC if CRD is the allocation mode and the GC interval is != 0. The default is non-zero (i.e. it can be disabled if you set the interval to 0)

tgraf commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,418 @@
+	identity := &v2.CiliumIdentity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   id.String(),
+			Labels: stripK8sPrefix(key.GetAsMap()),
```
I remember putting this in to get things going and start testing but for it to be reliable, do we instead need to translate `:` into another character? We can still get non `k8s:` prefixes, right? Which would then break the identity allocation.

raybejjani replied:
We do need to translate the `:`. I started translating it but when running in k8s I expected that they would always be `k8s:` sources. aanm suggested that we may also have `container:` sources. I'll check this again.

raybejjani replied:
Ok, from what I can tell, it is unlikely to get non-`k8s:` prefixed labels here. The problem is whether we care how the labels looks in k8s:
```
 $ kubectl get ciliumid 26636 -o json
{
    "apiVersion": "cilium.io/v2",
    "kind": "CiliumIdentity",
    "metadata": {
        "creationTimestamp": "2019-07-15T20:36:52Z",
        "generation": 1,
        "labels": {
            "class": "xwing",
            "io.cilium.k8s.policy.cluster": "default",
            "io.cilium.k8s.policy.serviceaccount": "default",
            "io.kubernetes.pod.namespace": "default",
            "org": "alliance"
        },
        "name": "26636",
        "resourceVersion": "10253",
        "selfLink": "/apis/cilium.io/v2/ciliumidentities/26636",
        "uid": "a6ddb730-8d80-4d6f-97ff-f6f03db393d9"
    },
    "status": {
        "nodes": {
            "192.168.34.11": "2019-07-15T21:23:54Z"
        }
    }
}
```

The only other separator character allowed, which I translated `:` before, is `_`. The labels would look like:
```
        "labels": {
            "k8s_class": "xwing",
            "k8s_io.cilium.k8s.policy.cluster": "default",
            "k8s_io.cilium.k8s.policy.serviceaccount": "default",
            "k8s_io.kubernetes.pod.namespace": "default",
            "k8s_org": "alliance"
        },
```

I'll think about this some more, but I'll leave it as is right now.

tgraf commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,418 @@
+	}
+
+	if capabilities.UpdateStatus {
+		_, err = identityOps.UpdateStatus(identityCopy.CiliumIdentity)
```
Do we need retries here? If Update fails because of conflicts, we will never release the identity.

raybejjani replied:
pkg/allocator.Allocator retries, but only for allocate. This includes calls AcquireReference. In theory, a failed Release will be garbage collected eventually. In any case, the kvstore backend also doesn't retry at that layer. Are you ok with slightly different behaviours between the backends?

raybejjani replied:
Discussed offline. I'll add some more comments/logging to ensure this case is clear, but pkg/allocator is the primary retry source.

joestringer commented on pkg/identity/cache/allocator.go:
```diff
@@ -145,9 +151,30 @@ func InitIdentityAllocator(owner IdentityAllocatorOwner) <-chan struct{} {
+			err     error
+		)
+
+		switch option.Config.IdentityAllocationMode {
```
This will segfault with a value other than `IdentityAllocationModeKVstore` or `IdentityAllocationModeCRD`.

aanm replied:
What's wrong with that?

joestringer replied:
It's a minor thing, but segfaulting (with stack trace) is not informative to users; it's more informative to actually print an error and `Fatal()`.

ianvernon replied:
I agree with @joestringer , `Fatal` indicates that it is a user configuration problem, while a panic / segfault indicates a code problem.

joestringer commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,418 @@
+			if err != nil {
+				return err
+			}
+			_, err = identityOps.Patch(identity.GetName(), k8sTypes.JSONPatchType, patch, "status")
```
If this error is non-nil, we will ignore it entirely right? Should we be logging it?

joestringer replied:
I presume this is the fallback case where we fall back to a full update of the CRD entry rather than patching it, so logically seems "ok".

raybejjani replied:
yes, it is a fallback but it is still an unexpected error. I'll debug print on error. That won't alarm anyone under normal usage, but might improve a debugger's life one day. A+1 question.

joestringer reviewed (commented):
A few fairly minor nits below. Plus I tried it out locally and hit:

```
level=error msg=k8sError error="github.com/cilium/cilium/pkg/k8s/identitybackend/identity.go:413: Failed to list *v2.CiliumIdentity: no kind \"CiliumIdentityList\" is registered for version \"cilium.io/v2\" in scheme \"github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/scheme/register.go:28\"" subsys=k8s
```

aanm commented on operator/k8s_identity.go:
```diff
@@ -0,0 +1,125 @@
+		&v2.CiliumIdentity{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
```
The cache will still receive the objects even if you don't have a handler for them. So you don't need the fake Add and Delete

aanm commented on operator/k8s_identity.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+func startCRDIdentityGC() {
+	for {
```
You mean a controller? ;-)

raybejjani replied:
I do! :P I'll switch it out

aanm commented on operator/main.go:
```diff
@@ -112,8 +114,10 @@ func init() {
 	flags.StringVar(&metricsAddress, "metrics-address", ":6942", "Address to serve Prometheus metrics")
 	flags.BoolVar(&synchronizeServices, "synchronize-k8s-services", true, "Synchronize Kubernetes services to kvstore")
 	flags.BoolVar(&synchronizeNodes, "synchronize-k8s-nodes", true, "Synchronize Kubernetes nodes to kvstore and perform CNP GC")
+	flags.DurationVar(&k8sIdentityHeartbeatTimeout, "identity-heartbeat-timeout", 15*time.Minute, "Timeout after which identity expires on lack of heartbeat")
```
Can't you import the lease timeout that we have for the nodes in the kvstore?

raybejjani replied:
good idea!

aanm commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,422 @@
+func stripK8sPrefix(old map[string]string) map[string]string {
+	fixup := make(map[string]string, len(old))
+	for k, v := range old {
+		k = strings.TrimPrefix(k, "k8s:")
```
The prefix `k8s` should be defined somewhere. I would create a helper to specific prefixes from labels?

raybejjani replied:
you're right, we should have it somewhere (without the `:` but that's ok). I thought the consensus was to use the labels inside the CRD as the real source of truth? The code I'm about to push up does 
```
k = strings.TrimPrefix(k, "k8s:")
k = strings.Replace(k, ":", "_", 0)
```

*but* those are just for the K8s meta labels, the real labels are inside the CRD

aanm commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,422 @@
+}
+
+// JSONPatch structure based on the RFC 6902
+type JSONPatch struct {
```
JSONPatch structure already exists, `pkg/k8s/json_patch.go`

raybejjani replied:
discussed offline. Sadly, this causes an import loop :/ We'll leave this in for now but it may be possible to solve this and a few other loops by changing `pkg/identity/cache`. I think it imports k8s/identitybackend only to initialise the identity backend.
```
package github.com/cilium/cilium/cilium
        imports github.com/cilium/cilium/cilium/cmd
        imports github.com/cilium/cilium/pkg/bpf
        imports github.com/cilium/cilium/pkg/k8s
        imports github.com/cilium/cilium/pkg/ipcache
        imports github.com/cilium/cilium/pkg/identity/cache
        imports github.com/cilium/cilium/pkg/k8s/identitybackend
        imports github.com/cilium/cilium/pkg/k8s
```

aanm reviewed (changes requested):
I'll review it later

aanm commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,422 @@
+		if err == nil {
+			return nil
+		}
+		log.WithError(err).Debug("Error patching status. This is not fatal")
```
You should invert the if statement above and log the error. I know the patching status is not fatal but we then should return if err != nil and not continue to execute the rest of the function.

aanm replied:
Oh I got it, you are continuing because you want to fallback to create the endpoint status with UpdateStatus and / or Update. Please add a comment describing that.

aanm commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,422 @@
+type CRDBackendConfiguration struct {
+	NodeName string
+	Store    cache.Store
+	Client   clientset.Interface
```
This could be the interface for the identity CRD (instead of being the interface of the full cilium CRD client)

raybejjani replied:
Discussed offline. ListAndWatch uses a `RESTClient` call and plumbing this around would be a hassle. We'll leave it as-is for now.

aanm commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,433 @@
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if identity, ok := obj.(*types.Identity); ok {
```
You need to check for `cache.DeletedFinalStateUnknown` object types. Like it is done here:

https://github.com/cilium/cilium/blob/23fd0d7cd4262ee1fc647f5a1078075b1f1f4fb0/daemon/k8s_watcher.go#L417-L430

raybejjani replied:
Done. It looks different from what we have elsewhere, so you might want to look at it again @aanm . I can add a `CopyObjToV1NetworkPolicy` style function to make it look similar. Up to you. I kept it simple for now.

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -458,6 +458,41 @@ type EndpointIdentity struct {
+	// SecurityLabels is the source-of-truth set of labels for this identity. The
+	// labels placed in metav1.ObjectMeta are there for convenience when
+	// searching and manipulating the CRD.
+	SecurityLabels map[string]string `json:"securityLabels"`
```
Sorry, I start looking at all the other structures that we have, it should be `json:"security-labels"` for consistency.

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -458,6 +458,41 @@ type EndpointIdentity struct {
+	// SecurityLabels is the source-of-truth set of labels for this identity. The
+	// labels placed in metav1.ObjectMeta are there for convenience when
+	// searching and manipulating the CRD.
+	SecurityLabels map[string]string `json:"securityLabels"`
```
Also, wayyy out of scope. Did you consider this to be `pkg/labels/labels.Labels` instead of a `map[string]string`? We can discuss offline.

raybejjani replied:
Hmm, I hadn't really. I guess it should work fine and might pretty-print better if we ever add that to labels.Labels. I'll try it out.

raybejjani replied:
ah, it does work but it looks a bit funky in json:
```
       {
            "apiVersion": "cilium.io/v2",
            "kind": "CiliumIdentity",
            "metadata": {
                "creationTimestamp": "2019-07-16T14:31:37Z",
                "generation": 1,
                "labels": {
                    "class": "deathstar",
                    "io.cilium.k8s.policy.cluster": "default",
                    "io.cilium.k8s.policy.serviceaccount": "default",
                    "io.kubernetes.pod.namespace": "default",
                    "org": "empire"
                },
                "name": "4508",
                "resourceVersion": "38127",
                "selfLink": "/apis/cilium.io/v2/ciliumidentities/4508",
                "uid": "d09ed5d6-35b4-4283-840e-a6832ffdf371"
            },
            "security-labels": {
                "class": {
                    "key": "class",
                    "source": "k8s",
                    "value": "deathstar"
                },
                "io.cilium.k8s.policy.cluster": {
                    "key": "io.cilium.k8s.policy.cluster",
                    "source": "k8s",
                    "value": "default"
                },
                "io.cilium.k8s.policy.serviceaccount": {
                    "key": "io.cilium.k8s.policy.serviceaccount",
                    "source": "k8s",
                    "value": "default"
                },
                "io.kubernetes.pod.namespace": {
                    "key": "io.kubernetes.pod.namespace",
                    "source": "k8s",
                    "value": "default"
                },
                "org": {
                    "key": "org",
                    "source": "k8s",
                    "value": "empire"
                }
            }
        }
```

aanm reviewed (changes requested):
Small stuff. Ping me offline if my comments were not clear

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -287,6 +287,32 @@ func createNodeCRD(clientset apiextensionsclient.Interface) error {
+func createIdentityCRD(clientset apiextensionsclient.Interface) error {
+	res := &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "ciliumidentities." + SchemeGroupVersion.Group,
```
Perhaps hardcode `ciliumidentities` and `ciliumidentity` to consts in case they are used elsewhere?

raybejjani replied:
ok, but 3/4 CRDs have the strings used inline. I switched it to be like the CEP creator. There are no global constants specific to the CRDs in this file.

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -458,6 +458,41 @@ type EndpointIdentity struct {
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CiliumIdentity is an identity managed by Cilium
```
Nit: this documentation is a bit too terse to me - perhaps refer to the identity structure in `pkg/identity`? Or even better, have some of what you put in the commit message here, since that describes its lifecycle / usage in a given Cilium deployment vs. just what it "is".

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -458,6 +458,41 @@ type EndpointIdentity struct {
+	metav1.ObjectMeta `json:"metadata"`
+
+	// SecurityLabels is the source-of-truth set of labels for this identity. The
+	// labels placed in metav1.ObjectMeta are there for convenience when
```
What exactly does "for convenience" mean in this case? Elaborate on whether the labels in `metav1.ObjectMeta` may be different than `SecurityLabels`.

ianvernon commented on operator/k8s_identity.go:
```diff
@@ -0,0 +1,116 @@
+	}
+
+nextIdentity:
+	for _, identityObject := range identityStore.List() {
```
Is it possible that after we do the `List` of an Identity has `len(identity.Status.Nodes) == 0`, that a K8s client in the cluster has then added the identity again? In other words, are we safe guarded against resource version changes for an identity after it has been listed here, but then updated shortly after `identityStore.List()` has been performed?

raybejjani replied:
It is guarded. `deleteIdentity` (called below) includes the `ResourceRevision` to the Delete call. That should return an error if the version doesn't match. It would log this and no-op. In the case of GC, this is benign and it will be retried in the next iteration.
Normally, identities would be deleted as soon as the final node removes itself. In those cases, the delete failing would leave the CRD instance for GC to clean-up. Both cases could race with new users of an identity. Generally, they would recreate the "master" key, in this case just the CRD, if an in-use identity is deleted. This is part of `pkg/allocator.Allocator`.

In any case, I added some more comments to try and make this more clear.`

ianvernon commented on operator/k8s_identity.go:
```diff
@@ -0,0 +1,116 @@
+	for _, identityObject := range identityStore.List() {
+		identity, ok := identityObject.(*types.Identity)
+		if !ok {
+			continue
```
Log an error here, this ideally should never happen. Refer to the type of the object as well.

ianvernon commented on pkg/identity/cache/allocator.go:
```diff
@@ -110,12 +113,15 @@ type IdentityAllocatorOwner interface {
+// currently, and identityStore may be nil.
+// Returns a channel which is closed when initialization of the allocator is
+// completed.
+func InitIdentityAllocator(owner IdentityAllocatorOwner, client clientset.Interface, identityStore cache.Store) <-chan struct{} {
```
I'm not a fan of pulling the K8s dependency into this package... ideally we'd be able to hide it behind an interface again, but that's a lot of work. Can you put a TODO here for this?

ianvernon commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,433 @@
+// kubernetes does not allow ':' in the name of the label. These labels are not
+// the canonical labels of the identity, but used to ease interaction with the
+// CRD object.
+func sanitizeK8sLabels(old map[string]string) map[string]string {
```
Please add a unit test for this.

ianvernon commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,433 @@
+	fixup := make(map[string]string, len(old))
+	for k, v := range old {
+		if !strings.HasPrefix(k, k8sPrefix) {
+			continue // skip non-k8s labels
```
Perhaps log that we are skipping a non-k8s label? It would be useful to warn to users / signal to them somehow that we are skipping labels.

raybejjani replied:
What would you say here to avoid confusion? The meta labels are the ones that drop non-k8s labels. The SecurityLabels field would hold all. I'm not sure how to convey that a label being ignored does not mean that it isn't used in the identity.

ianvernon replied:
Hm, maybe say that these labels are skipped only when we are adding them to the ObjectMeta, which does not mean we are removing them from the labels (`SecurityLabels`) used in the actual identity itself.

ianvernon commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,433 @@
+		if err == nil {
+			return nil
+		}
+		log.WithError(err).Debug("Error updating status. This Continuing update via Update")
```
This --> Thus ?

raybejjani replied:
oops! that was a left over from when that said "This is not fatal". There are 4 messages of this sort, 2 at each "step" in an update

ianvernon commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,433 @@
+	if reliablyMissing {
+		// Recreate a missing master key
+		if err = c.AllocateID(ctx, id, key); err != nil {
+			return fmt.Errorf("Unable recreate missing CRD identity \"%s\"->\"%s\": %s", key, id, err)
```
You can use `%q` instead of `\"%s\"`

ianvernon commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -0,0 +1,433 @@
+	c.Store = cache.NewStore(cache.DeletionHandlingMetaNamespaceKeyFunc)
+	identityInformer := informer.NewInformerWithStore(
+		cache.NewListWatchFromClient(c.Client.CiliumV2().RESTClient(),
+			"ciliumidentities", v1.NamespaceAll, fields.Everything()),
```
Use the aforementioned constant I asked for earlier.

ianvernon commented on pkg/option/config.go:
```diff
@@ -1654,6 +1671,17 @@ func (c *DaemonConfig) Populate() {
+	case c.IdentityAllocationMode == "":
+		c.IdentityAllocationMode = IdentityAllocationModeKVstore
+
+	case c.IdentityAllocationMode != IdentityAllocationModeKVstore &&
```
Nit: why not have, 
```
case IdentityAllocationModeKVStore, caseIdentityAllocationModeCRD:
default:
    log.Fatal(
...
```

It would be easier to parse / more in line with how we handle invalid config in other parts of the code.
